File: android/src/main/java/cn/jjvu/xiao/easy_flutter_amap/EasyFlutterAmapPlugin.java
package cn.jjvu.xiao.easy_flutter_amap;

import android.app.Activity;

import androidx.annotation.NonNull;

import cn.jjvu.xiao.easy_flutter_amap.view.AmapViewFactory;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.platform.PlatformViewRegistry;

public class EasyFlutterAmapPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware {

  private MethodChannel channel;
  
  private BinaryMessenger messenger;
  
  private PlatformViewRegistry platformViewRegistry;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
//    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "easy_flutter_amap");
//    channel.setMethodCallHandler(this);
    messenger = flutterPluginBinding.getBinaryMessenger();;
    platformViewRegistry = flutterPluginBinding.getPlatformViewRegistry();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
    Activity activity = binding.getActivity();
    platformViewRegistry.registerViewFactory("cn.jjvu.xiao.easy_flutter_amap/mapview", new AmapViewFactory(messenger, activity));
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {

  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

  }

  @Override
  public void onDetachedFromActivity() {

  }
}

File: android/src/main/kotlin/cn/jjvu/xiao/easy_flutter_amap/EasyFlutterAmapPlugin.kt
package cn.jjvu.xiao.easy_flutter_amap

import androidx.annotation.NonNull
import com.amap.api.maps.model.LatLng
import io.flutter.Log
import io.flutter.embedding.android.FlutterActivity

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlin.math.log

/** EasyFlutterAmapPlugin */
class EasyFlutterAmapPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var _activity: FlutterActivity

    private lateinit var _flutterPluginBinding: FlutterPlugin.FlutterPluginBinding

    private lateinit var _channel: MethodChannel

    private lateinit var _aMapViewFactory: AMapViewFactory

    private val TAG = "easy_flutter_amap"

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        this._flutterPluginBinding = flutterPluginBinding
        _channel = MethodChannel(flutterPluginBinding.binaryMessenger, TAG)
        _channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        _channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when(call.method) {
            "setMyLocationType" -> _aMapViewFactory.aMapView.setLocationType(call.arguments<Int>())

        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        _activity = binding.activity as FlutterActivity
        _aMapViewFactory = AMapViewFactory(_activity, _flutterPluginBinding.binaryMessenger)
        _flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "AMapView",
            _aMapViewFactory
        )
    }

    override fun onDetachedFromActivityForConfigChanges() {

    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {

    }

    override fun onDetachedFromActivity() {

    }
}
